File: bindings/kotlin/src/main/kotlin/at/schem/nucleation/PaletteBuilder.kt
package at.schem.nucleation;
import com.sun.jna.Callback
import com.sun.jna.Library
import com.sun.jna.Native
import com.sun.jna.Pointer
import com.sun.jna.Structure

internal interface PaletteBuilderLib: Library {
    fun PaletteBuilder_destroy(handle: Pointer)
    fun PaletteBuilder_create(): Pointer
    fun PaletteBuilder_exclude_falling(handle: Pointer): ResultUnitInt
    fun PaletteBuilder_exclude_tile_entities(handle: Pointer): ResultUnitInt
    fun PaletteBuilder_full_blocks_only(handle: Pointer): ResultUnitInt
    fun PaletteBuilder_exclude_needs_support(handle: Pointer): ResultUnitInt
    fun PaletteBuilder_exclude_transparent(handle: Pointer): ResultUnitInt
    fun PaletteBuilder_exclude_light_sources(handle: Pointer): ResultUnitInt
    fun PaletteBuilder_survival_only(handle: Pointer): ResultUnitInt
    fun PaletteBuilder_exclude_keyword(handle: Pointer, keyword: Slice): ResultUnitInt
    fun PaletteBuilder_include_keyword(handle: Pointer, keyword: Slice): ResultUnitInt
    fun PaletteBuilder_build(handle: Pointer): ResultPointerInt
}
/** Filter-driven palette construction (wraps
*[crate::building::PaletteBuilder], which fronts blockpedia's
*`BlockFilter`). Call flag methods, then `build` — the builder is
*consumed; further calls error with `AlreadyConsumed`.
*/
class PaletteBuilder internal constructor (
    internal val handle: Pointer,
    // These ensure that anything that is borrowed is kept alive and not cleaned
    // up by the garbage collector.
    internal val selfEdges: List<Any>,
    internal var owned: Boolean,
)  {

    init {
        if (this.owned) {
            this.registerCleaner()
        }
    }

    private class PaletteBuilderCleaner(val handle: Pointer, val lib: PaletteBuilderLib) : Runnable {
        override fun run() {
            lib.PaletteBuilder_destroy(handle)
        }
    }
    private fun registerCleaner() {
        CLEANER.register(this, PaletteBuilder.PaletteBuilderCleaner(handle, PaletteBuilder.lib));
    }

    companion object {
        internal val libClass: Class<PaletteBuilderLib> = PaletteBuilderLib::class.java
        internal val lib: PaletteBuilderLib = Native.load("nucleation", libClass)
        @JvmStatic
        
        /** A builder matching every colored block (no filters yet).
        */
        fun create(): PaletteBuilder {
            
            val returnVal = lib.PaletteBuilder_create();
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = PaletteBuilder(handle, selfEdges, true)
            return returnOpaque
        }
    }
    
    /** Exclude gravity-affected blocks (sand, gravel, ...).
    */
    fun excludeFalling(): Result<Unit> {
        
        val returnVal = lib.PaletteBuilder_exclude_falling(handle);
        val nativeOkVal = returnVal.getNativeOk();
        if (nativeOkVal != null) {
            return Unit.ok()
        } else {
            return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
        }
    }
    
    /** Exclude blocks with block entities (chests, furnaces, ...).
    */
    fun excludeTileEntities(): Result<Unit> {
        
        val returnVal = lib.PaletteBuilder_exclude_tile_entities(handle);
        val nativeOkVal = returnVal.getNativeOk();
        if (nativeOkVal != null) {
            return Unit.ok()
        } else {
            return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
        }
    }
    
    /** Keep only full cube blocks (no stairs, slabs, fences, ...).
    */
    fun fullBlocksOnly(): Result<Unit> {
        
        val returnVal = lib.PaletteBuilder_full_blocks_only(handle);
        val nativeOkVal = returnVal.getNativeOk();
        if (nativeOkVal != null) {
            return Unit.ok()
        } else {
            return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
        }
    }
    
    /** Exclude blocks that need supporting blocks (torches, rails, ...).
    */
    fun excludeNeedsSupport(): Result<Unit> {
        
        val returnVal = lib.PaletteBuilder_exclude_needs_support(handle);
        val nativeOkVal = returnVal.getNativeOk();
        if (nativeOkVal != null) {
            return Unit.ok()
        } else {
            return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
        }
    }
    
    /** Exclude transparent/translucent blocks (glass, leaves, ...).
    */
    fun excludeTransparent(): Result<Unit> {
        
        val returnVal = lib.PaletteBuilder_exclude_transparent(handle);
        val nativeOkVal = returnVal.getNativeOk();
        if (nativeOkVal != null) {
            return Unit.ok()
        } else {
            return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
        }
    }
    
    /** Exclude light-emitting blocks (glowstone, lanterns, ...).
    */
    fun excludeLightSources(): Result<Unit> {
        
        val returnVal = lib.PaletteBuilder_exclude_light_sources(handle);
        val nativeOkVal = returnVal.getNativeOk();
        if (nativeOkVal != null) {
            return Unit.ok()
        } else {
            return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
        }
    }
    
    /** Keep only blocks obtainable in survival.
    */
    fun survivalOnly(): Result<Unit> {
        
        val returnVal = lib.PaletteBuilder_survival_only(handle);
        val nativeOkVal = returnVal.getNativeOk();
        if (nativeOkVal != null) {
            return Unit.ok()
        } else {
            return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
        }
    }
    
    /** Exclude blocks whose id contains `keyword`.
    */
    fun excludeKeyword(keyword: String): Result<Unit> {
        val keywordSliceMemory = PrimitiveArrayTools.borrowUtf8(keyword)
        
        val returnVal = lib.PaletteBuilder_exclude_keyword(handle, keywordSliceMemory.slice);
        try {
            val nativeOkVal = returnVal.getNativeOk();
            if (nativeOkVal != null) {
                return Unit.ok()
            } else {
                return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
            }
        } finally {
            keywordSliceMemory.close()
        }
    }
    
    /** Keep only blocks whose id contains `keyword` (repeatable; matches
    *any of the included keywords).
    */
    fun includeKeyword(keyword: String): Result<Unit> {
        val keywordSliceMemory = PrimitiveArrayTools.borrowUtf8(keyword)
        
        val returnVal = lib.PaletteBuilder_include_keyword(handle, keywordSliceMemory.slice);
        try {
            val nativeOkVal = returnVal.getNativeOk();
            if (nativeOkVal != null) {
                return Unit.ok()
            } else {
                return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
            }
        } finally {
            keywordSliceMemory.close()
        }
    }
    
    /** Build the palette; consumes the builder.
    */
    fun build(): Result<Palette> {
        
        val returnVal = lib.PaletteBuilder_build(handle);
        val nativeOkVal = returnVal.getNativeOk();
        if (nativeOkVal != null) {
            val selfEdges: List<Any> = listOf()
            val handle = nativeOkVal 
            val returnOpaque = Palette(handle, selfEdges, true)
            return returnOpaque.ok()
        } else {
            return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
        }
    }

}

File: bindings/php/PaletteBuilder.php
<?php
namespace Stencil;

final class PaletteBuilder {
    /** @internal */
    public \FFI\CData $ptr;
    private bool $owned;
    private ?object $borrowedFrom;

    /** @internal */
    public function __construct(\FFI\CData $ptr, bool $owned, ?object $borrowedFrom = null) {
        $this->ptr = $ptr;
        $this->owned = $owned;
        $this->borrowedFrom = $borrowedFrom;
    }

    public static function create() {
        $ret = Lib::ffi()->PaletteBuilder_create();
        return new PaletteBuilder($ret, true);
    }

    public function excludeFalling() {
        $result = Lib::ffi()->PaletteBuilder_exclude_falling($this->ptr);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
    }

    public function excludeTileEntities() {
        $result = Lib::ffi()->PaletteBuilder_exclude_tile_entities($this->ptr);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
    }

    public function fullBlocksOnly() {
        $result = Lib::ffi()->PaletteBuilder_full_blocks_only($this->ptr);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
    }

    public function excludeNeedsSupport() {
        $result = Lib::ffi()->PaletteBuilder_exclude_needs_support($this->ptr);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
    }

    public function excludeTransparent() {
        $result = Lib::ffi()->PaletteBuilder_exclude_transparent($this->ptr);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
    }

    public function excludeLightSources() {
        $result = Lib::ffi()->PaletteBuilder_exclude_light_sources($this->ptr);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
    }

    public function survivalOnly() {
        $result = Lib::ffi()->PaletteBuilder_survival_only($this->ptr);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
    }

    public function excludeKeyword(string $keyword) {
        $__n0 = strlen($keyword);
        $__view0 = Lib::ffi()->new('DiplomatStringView');
        if ($__n0 > 0) {
            $__buf0 = Lib::ffi()->new("uint8_t[" . $__n0 . "]", false);
            \FFI::memcpy($__buf0, $keyword, $__n0);
            $__view0->data = Lib::ffi()->cast('const char*', \FFI::addr($__buf0[0]));
        } else {
            $__view0->data = null;
        }
        $__view0->len = $__n0;
        $result = Lib::ffi()->PaletteBuilder_exclude_keyword($this->ptr, $__view0);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
    }

    public function includeKeyword(string $keyword) {
        $__n0 = strlen($keyword);
        $__view0 = Lib::ffi()->new('DiplomatStringView');
        if ($__n0 > 0) {
            $__buf0 = Lib::ffi()->new("uint8_t[" . $__n0 . "]", false);
            \FFI::memcpy($__buf0, $keyword, $__n0);
            $__view0->data = Lib::ffi()->cast('const char*', \FFI::addr($__buf0[0]));
        } else {
            $__view0->data = null;
        }
        $__view0->len = $__n0;
        $result = Lib::ffi()->PaletteBuilder_include_keyword($this->ptr, $__view0);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
    }

    public function tag(string $t) {
        $__n0 = strlen($t);
        $__view0 = Lib::ffi()->new('DiplomatStringView');
        if ($__n0 > 0) {
            $__buf0 = Lib::ffi()->new("uint8_t[" . $__n0 . "]", false);
            \FFI::memcpy($__buf0, $t, $__n0);
            $__view0->data = Lib::ffi()->cast('const char*', \FFI::addr($__buf0[0]));
        } else {
            $__view0->data = null;
        }
        $__view0->len = $__n0;
        $result = Lib::ffi()->PaletteBuilder_tag($this->ptr, $__view0);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
    }

    public function excludeTag(string $t) {
        $__n0 = strlen($t);
        $__view0 = Lib::ffi()->new('DiplomatStringView');
        if ($__n0 > 0) {
            $__buf0 = Lib::ffi()->new("uint8_t[" . $__n0 . "]", false);
            \FFI::memcpy($__buf0, $t, $__n0);
            $__view0->data = Lib::ffi()->cast('const char*', \FFI::addr($__buf0[0]));
        } else {
            $__view0->data = null;
        }
        $__view0->len = $__n0;
        $result = Lib::ffi()->PaletteBuilder_exclude_tag($this->ptr, $__view0);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
    }

    public function kind(string $k) {
        $__n0 = strlen($k);
        $__view0 = Lib::ffi()->new('DiplomatStringView');
        if ($__n0 > 0) {
            $__buf0 = Lib::ffi()->new("uint8_t[" . $__n0 . "]", false);
            \FFI::memcpy($__buf0, $k, $__n0);
            $__view0->data = Lib::ffi()->cast('const char*', \FFI::addr($__buf0[0]));
        } else {
            $__view0->data = null;
        }
        $__view0->len = $__n0;
        $result = Lib::ffi()->PaletteBuilder_kind($this->ptr, $__view0);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
    }

    public function lightnessBetween( $min,  $max) {
        $result = Lib::ffi()->PaletteBuilder_lightness_between($this->ptr, $min, $max);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
    }

    public function chromaBelow( $max) {
        $result = Lib::ffi()->PaletteBuilder_chroma_below($this->ptr, $max);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
    }

    public function colorNear( $r,  $g,  $b,  $max_distance) {
        $result = Lib::ffi()->PaletteBuilder_color_near($this->ptr, $r, $g, $b, $max_distance);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
    }

    public function build() {
        $result = Lib::ffi()->PaletteBuilder_build($this->ptr);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
        return new Palette($result->ok, true);
    }

    public function __destruct() {
        if ($this->owned) {
            Lib::ffi()->PaletteBuilder_destroy($this->ptr);
        }
    }
}
